modules/gmake/src/proj.lua

```diff
@@ -0,0 +1,677 @@
+-- @param prj
+--  project to print object directory rule of
+---
+function proj.objDirRule(prj)
+	wl('$(OBJDIR):')
+	export.indent()
+	wl('@echo "Creating $(OBJDIR)"')
+	export.outdent()
+	wl('ifeq (posix,$(SHELLTYPE))')
+	export.indent()
+	wl('$(SILENT) mkdir -p $(OBJDIR)')
+	export.outdent()
+	wl('else')
+	export.indent()
+	wl('$(SILENT) mkdir $(subst /,\\\\,$(OBJDIR))')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the target directory rule.
+--
+-- @param prj
+--  project to print target directory rule of
+---
+function proj.targetDirRule(prj)
+	wl('$(TARGETDIR):')
+	export.indent()
+	wl('@echo "Creating $(TARGETDIR)"')
+	export.outdent()
+	wl('ifeq (posix,$(SHELLTYPE))')
+	export.indent()
+	wl('$(SILENT) mkdir -p $(TARGETDIR)')
+	export.outdent()
+	wl('else')
+	export.indent()
+	wl('$(SILENT) mkdir $(subst /,\\\\,$(TARGETDIR))')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the target rule.
+--
+-- @param prj
+--  project to print target rule of
+---
+function proj.targetRule(prj)
+	wl('$(TARGET): $(OBJECTS) | $(TARGETDIR)')
+	export.indent()
+	wl('$(PRELINKCMDS)')
+	wl('@echo "Linking %s"', prj.name)
+	wl('$(SILENT) $(LINKCMD)')
+	wl('$(POSTBUILDCMDS)')
+	export.outdent()
+	wl()
+end
+
+
+---
+-- Prints the clean rule.
+--
+-- @param prj
+--  project to print clean rule of
+---
+function proj.cleanRule(prj)
+	wl('clean:')
+	export.indent()
+	wl('@echo "Cleaning %s"', prj.name)
+	export.outdent()
+	wl('ifeq (posix,$(SHELLTYPE))')
+	export.indent()
+	wl('$(SILENT) rm -f  $(TARGET)')
+	wl('$(SILENT) rm -rf $(OBJDIR)')
+	export.outdent()
+	wl('else')
+	export.indent()
+	wl('$(SILENT) if exist $(subst /,\\\\,$(TARGET)) del $(subst /,\\\\,$(TARGET))')
+	wl('$(SILENT) if exist $(subst /,\\\\,$(OBJDIR)) rmdir /s /q $(subst /,\\\\,$(OBJDIR))')
+	export.outdent()
+	wl('endif')
+	wl()
+end
+
+
+---
+-- Prints the all rule of the project.
+--
+-- @param prj
+--  project to print all rule of
+---
+function proj.allRule(prj)
+	wl('all: $(TARGET)')
+	export.indent()
+	wl('@:')
+	export.outdent()
+	wl()
+end
+
+
+---
+-- Prints the object file definition.
+--
+-- @param prj
+--  Project to print object files of
+---
+function proj.objects(prj)
+	wl('OBJECTS :=')
+	wl()
+
+	local files = {}
+
+	for _, file in ipairs(prj.files) do
+		if string.endsWith(file, proj.SOURCE_FILES) then
+			wl('OBJECTS += $(OBJDIR)/%s.o', path.getBaseName(file))
+		end
+	end
+	wl()
+end
+
+
+---
+-- File configuration rules.
+--
+-- @param prj
+--  project to print file rules for
+---
+function proj.fileRules(prj)
+	wl('# File Rules')
+	wl()
+
+	for _, file in ipairs(prj.files) do
+		if string.endsWith(file, proj.SOURCE_FILES) then
+			wl('$(OBJDIR)/%s.o: %s', path.getBaseName(file), path.getRelative(prj.location, file))
+			export.indent()
+			wl('@echo $(notdir $<)')
+			-- TODO: This should be $(CC) for C builds
+			wl('$(SILENT) $(CXX) $(ALL_CXXFLAGS) $(FORCE_INCLUDE) -o "$@" -MF "$(@:%%.o=%%.d)" -c "$<"')
+			export.outdent()
+			wl()
```

Comment:
To make overriding easier, can we break out the per file-type logic into their own methods? Maybe like…

```lua
if string.endsWith(file, prod.SOURCE_FILES) then
   proj.sourceFileRule(prj, file)
end
```